array.c

```diff
@@ -3232,6 +3233,9 @@ sort_1(const void *ap, const void *bp, void *dummy)
     args[0] = a;
     args[1] = b;
     retval = rb_yield_values2(2, args);
+    if (rb_obj_frozen_p(data->receiver)) {
+        rb_raise(rb_eFrozenError, "array frozen during sort");
+    }
     n = rb_cmpint(retval, a, b);
```

Comment:
Also `retval.<` or `retval.>` in `rb_cmpint` can freeze the receiver.
And `sort_2` needs this check too.

Reply:
@nobu  Thank you for your review.  I have rebased this on master and fixed the two cases you mentioned.

Reply:
Maybe extract a function with `sort_reentered`?

Reply:
That makes sense.  I've pushed another commit that adds a `sort_returned` function that calls `sort_reentered` and has the frozen check.